ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs

```diff
@@ -434,6 +448,187 @@ traceLedgerProcessing
+                            (tHdrTick  `div` 1000)
+                            (tHdrApp   `div` 1000)
+                            (tBlkTick  `div` 1000)
+                            (tBlkApp   `div` 1000)
+          -- Compute the difference between two 'RtsTime's and convert it to microseconds.
+          nsDiff t t' = (t - t') `div` 1000
+
+          slotCount (SlotNo i) = \case
+            Slotting.Origin        -> i
+            Slotting.At (SlotNo j) -> i - j
+
+        IO.hPutStrLn outFileHandle (   DP.showData slotDataPoint separator
+                                    ++ intercalate separator (HasAnalysis.blockStats blk)
+                                   )
+
+        pure $ BenchmarkLedgerOpsState currentRtsStats $ ExtLedgerState ldgrSt' hdrSt'
+      where
+        rp = blockRealPoint blk
+
+        forecast ::
+             SlotNo
+          -> ExtLedgerState blk
+          -> IO (Ticked (LedgerView (BlockProtocol blk)), GC.RtsTime)
+        forecast slot st = time chrono $ do
+          let forecaster = ledgerViewForecastAt lcfg (ledgerState st)
+          case runExcept $ forecastFor forecaster slot of
+            Left err -> fail $ "benchmark doesn't support headers beyond the forecast limit: " <> show rp <> " " <> show err
+            Right !x -> pure x
+
+        tickTheHeaderState ::
+             SlotNo
+          -> ExtLedgerState blk
+          -> Ticked (LedgerView (BlockProtocol blk))
+          -> IO (Ticked (HeaderState blk), GC.RtsTime)
+        tickTheHeaderState slot st tickedLedgerView = time chrono $
+          pure $! tickHeaderState ccfg
+                                  tickedLedgerView
+                                  slot
+                                  (headerState st)
+
+        applyTheHeader ::
+             Ticked (LedgerView (BlockProtocol blk))
+          -> Ticked (HeaderState blk)
+          -> IO (HeaderState blk, GC.RtsTime)
+        applyTheHeader tickedLedgerView tickedHeaderState = time chrono $ do
+          case runExcept $ validateHeader cfg tickedLedgerView (getHeader blk) tickedHeaderState of
+            Left err -> fail $ "benchmark doesn't support invalid headers: " <> show rp <> " " <> show err
+            Right x -> pure x
+
+        tickTheLedgerState ::
+             SlotNo
+          -> ExtLedgerState blk
+          -> IO (Ticked (LedgerState blk), GC.RtsTime)
+        tickTheLedgerState slot st =
+          time chrono $! pure $ applyChainTick lcfg slot (ledgerState st)
+
+        applyTheBlock ::
+             Ticked (LedgerState blk)
+          -> IO (LedgerState blk, GC.RtsTime)
+        applyTheBlock tickedLedgerSt = time chrono $ do
+          case runExcept (lrResult <$> applyBlockLedgerResult lcfg blk tickedLedgerSt) of
+            Left err -> fail $ "benchmark doesn't support invalid blocks: " <> show rp <> " " <> show err
+            Right x -> pure x
+
+newtype Chronometer = Chronometer { mutatorElapsedTimeNS :: IORef GC.RtsTime }
```

Comment:
You could use a `StateT IO` stack if you don't want to go this "objecty" route.

Reply:
Good point. I don't have any preference, but I wonder if we need this IORef at all, isn't the cost of the extra `getRTSStats` negligible when compared with the other ledger ops? I'd love to get rid of the `Chronometer` :stopwatch:

Reply:
I would guess it's relatively very low-cost. Moreover, "one extra" doesn't matter: we'll be calculating the difference between them, so the last one's cost is skipped (in the measurements, but not in the analysis' actual run-time).

Reply:
Yeah, sounds sensisble. I removed it.